alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,108 @@
+use std::cmp;
+
+use glutin::Rect;
+
+use alacritty_terminal::term::{LineDamageInformation, SizeInfo, TermDamageIterator};
+
+/// Minimum part of damaged area when merging damaged rectangles.
+const MIN_DAMAGED_AREA_PART: f32 = 0.4;
+
+/// Error of minimum part of damaged area in which we try to merge more rectangles until we reach
+/// [`MIN_DAMAGED_AREA_PART`].
+const MIN_DAMAGED_AREA_PART_ERR: f32 = MIN_DAMAGED_AREA_PART / 2.;
+
+/// Iterator which converts `alacritty_terminal` damage information into renderer damaged rects.
+pub struct RenderDamageIterator<'a> {
+    damaged_lines: TermDamageIterator<'a>,
+    size_info: SizeInfo<u32>,
+}
+
+impl<'a> RenderDamageIterator<'a> {
+    pub fn new(damaged_lines: TermDamageIterator<'a>, size_info: SizeInfo<u32>) -> Self {
+        Self { damaged_lines, size_info }
+    }
+
+    #[inline]
+    fn rect_for_line(&self, line_damage: LineDamageInformation) -> Rect {
+        let size_info = &self.size_info;
+        let y_top = size_info.height() - size_info.padding_y();
+        let x = size_info.padding_x() + line_damage.left as u32 * size_info.cell_width();
+        let y = y_top - (line_damage.line + 1) as u32 * size_info.cell_height();
+        let width = (line_damage.right - line_damage.left + 1) as u32 * size_info.cell_width();
+        Rect { x, y, height: size_info.cell_height(), width }
+    }
+
+    // Make sure to damage near cells to include wide chars.
+    #[inline]
+    fn overdamage(&self, mut rect: Rect) -> Rect {
+        let size_info = &self.size_info;
+        rect.x = rect.x.saturating_sub(size_info.cell_width());
+        rect.width = cmp::min(size_info.width() - rect.x, rect.width + 2 * size_info.cell_width());
+        rect.y = rect.y.saturating_sub(size_info.cell_height() / 2);
+        rect.height = cmp::min(size_info.height() - rect.y, rect.height + size_info.cell_height());
+
+        rect
+    }
+}
+
+impl<'a> Iterator for RenderDamageIterator<'a> {
+    type Item = Rect;
+
+    fn next(&mut self) -> Option<Rect> {
+        let line = self.damaged_lines.next()?;
+
+        let mut rect = self.rect_for_line(line);
+        let mut advance = 0;
+
+        let mut ahead_rect = rect;
+        let mut ahead_area = rect_area(&ahead_rect);
+
+        // The amount we advanced by from the last picked rect.
+        let mut advanced_ahead = 0;
+
+        for line in self.damaged_lines.clone() {
```

Comment:
There's no need to and it won't work since we're mutating the iterator and not touching the original until we ensure that we should. Cloning iterator is cheap anyway.